File: ts/src/lib/emitters/movie.ts
import { WritableByteStream, WritableStream } from "@open-flash/stream";
import { Incident } from "incident";
import { UintSize } from "semantic-types";
import { CompressionMethod, Header, Movie, SwfSignature } from "swf-tree";
import * as zlib from "zlib";
import { emitRect } from "./basic-data-types";
import { emitTagString } from "./tags";

const SWF_SIGNATURE_SIZE: UintSize = 8;

export function emitMovie(byteStream: WritableByteStream, value: Movie, compressionMethod: CompressionMethod): void {
  const movieStream: WritableByteStream = new WritableStream();
  emitPayload(movieStream, value);
  const uncompressedFileLength: UintSize = SWF_SIGNATURE_SIZE + movieStream.bytePos;
  const signature: SwfSignature = {
    compressionMethod,
    swfVersion: value.header.swfVersion,
    uncompressedFileLength,
  };
  emitSwfSignature(byteStream, signature);
  switch (compressionMethod) {
    case CompressionMethod.Deflate:
      const movieBytes: Uint8Array = movieStream.getBytes();
      byteStream.writeBytes(zlib.deflateSync(<any> movieBytes));
      break;
    case CompressionMethod.Lzma:
      throw new Incident("NotImplemented", "Lzma support");
    case CompressionMethod.None:
      byteStream.write(movieStream);
      break;
    default:
      throw new Incident("UnexpectedCompressionMethod");
  }
}

export function emitSwfSignature(byteStream: WritableByteStream, value: SwfSignature): void {
  emitCompressionMethod(byteStream, value.compressionMethod);
  byteStream.writeUint8(value.swfVersion);
  byteStream.writeUint32LE(value.uncompressedFileLength);
}

export function emitCompressionMethod(byteStream: WritableByteStream, value: CompressionMethod): void {
  const COMPRESSION_TO_CHUNK: Map<CompressionMethod, Uint8Array> = new Map([
    [CompressionMethod.Deflate, new Uint8Array([0x43, 0x57, 0x53])],
    [CompressionMethod.Lzma, new Uint8Array([0x5a, 0x57, 0x53])],
    [CompressionMethod.None, new Uint8Array([0x46, 0x57, 0x53])],
  ]);

  const chunk: Uint8Array | undefined = COMPRESSION_TO_CHUNK.get(value);
  if (chunk === undefined) {
    throw new Incident("UnexpectedCompressionMethod");
  }
  byteStream.writeBytes(chunk);
}

function emitPayload(byteStream: WritableByteStream, value: Movie): void {
  emitHeader(byteStream, value.header);
  emitTagString(byteStream, value.tags, value.header.swfVersion);
}

export function emitHeader(byteStream: WritableByteStream, value: Header): void {
  emitRect(byteStream, value.frameSize);
  byteStream.writeUint16LE(value.frameRate.epsilons);
  byteStream.writeUint16LE(value.frameCount);
}

File: rs/src/movie.rs
use std::convert::TryInto;
use std::io;

use crate::basic_data_types::emit_rect;
use crate::error::SwfEmitError;
use crate::primitives::{emit_le_u16, emit_le_u32, emit_u8};
use crate::tags::emit_tag_string;
use swf_types as ast;

const SWF_SIGNATURE_SIZE: usize = 8;

pub fn emit_swf<W: io::Write>(
  writer: &mut W,
  value: &ast::Movie,
  compression_method: ast::CompressionMethod,
) -> Result<(), SwfEmitError> {
  let write_movie_fn = match compression_method {
    ast::CompressionMethod::None => W::write_all,
    #[cfg(feature="deflate")]
    ast::CompressionMethod::Deflate => write_bytes_deflate,
    #[cfg(feature="lzma")]
    ast::CompressionMethod::Lzma => write_bytes_lzma,
    #[allow(unreachable_patterns)]
    method => return Err(SwfEmitError::UnsupportedCompression(method)),
  };

  let mut movie_bytes = Vec::new();
  emit_movie(&mut movie_bytes, value).map_err(SwfEmitError::Io)?;
  let uncompressed_file_length = SWF_SIGNATURE_SIZE + movie_bytes.len();
  let signature = ast::SwfSignature {
    compression_method,
    swf_version: value.header.swf_version,
    uncompressed_file_length,
  };

  emit_swf_signature(writer, &signature).map_err(SwfEmitError::Io)?;
  write_movie_fn(writer, &movie_bytes).map_err(SwfEmitError::Io)
}

#[cfg(feature="lzma")]
fn write_bytes_lzma<W: io::Write>(writer: &mut W, mut bytes: &[u8]) -> io::Result<()> {
  lzma_rs::lzma_compress(&mut bytes, writer)
}

#[cfg(feature="deflate")]
fn write_bytes_deflate<W: io::Write>(writer: &mut W, bytes: &[u8]) -> io::Result<()> {
  use miniz_oxide::deflate::{core, CompressionLevel};

  let mut compressor = core::CompressorOxide::default();
  compressor.set_format_and_level(miniz_oxide::DataFormat::Zlib, CompressionLevel::DefaultLevel as u8);

  let mut result = Ok(());
  let (status, written) = core::compress_to_output(&mut compressor, bytes, core::TDEFLFlush::Finish, |out| {
    result = writer.write_all(out);
    result.is_ok()
  });
  
  // Check the compression status; panicking here is a bug.
  match status {
    core::TDEFLStatus::PutBufFailed => assert!(result.is_err(), "miniz_oxide: unexpected error"),
    core::TDEFLStatus::BadParam => panic!("miniz_oxide: bad deflate params"),
    core::TDEFLStatus::Okay => panic!("miniz_oxide: unexpected partial compression"),
    core::TDEFLStatus::Done => assert!(result.is_ok() && written == bytes.len(), "miniz_oxide: unexpected success"),
  }
  result
}

pub fn emit_swf_signature<W: io::Write>(writer: &mut W, value: &ast::SwfSignature) -> io::Result<()> {
  assert!(value.uncompressed_file_length <= 0xffffffff);

  emit_compression_method(writer, value.compression_method)?;
  emit_u8(writer, value.swf_version)?;
  emit_le_u32(writer, value.uncompressed_file_length.try_into().unwrap())
}

pub fn emit_compression_method<W: io::Write>(writer: &mut W, value: ast::CompressionMethod) -> io::Result<()> {
  let code: &'static [u8] = match value {
    ast::CompressionMethod::Deflate => b"CWS",
    ast::CompressionMethod::Lzma => b"ZWS",
    ast::CompressionMethod::None => b"FWS",
  };
  writer.write_all(code)
}

pub fn emit_movie<W: io::Write>(writer: &mut W, value: &ast::Movie) -> io::Result<()> {
  emit_header(writer, &value.header)?;
  emit_tag_string(writer, &value.tags, value.header.swf_version)
}

pub fn emit_header<W: io::Write>(writer: &mut W, value: &ast::Header) -> io::Result<()> {
  emit_rect(writer, &value.frame_size)?;
  emit_le_u16(writer, value.frame_rate.epsilons)?;
  emit_le_u16(writer, value.frame_count)
}
